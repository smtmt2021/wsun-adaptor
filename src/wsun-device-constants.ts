/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export const MAX_BUF_SIZE = 1024;
export const MAX_DGRAM_SIZE = 1024;

const isTesting = process.env.NODE_ENV === 'test';
export const COMMAND_TIMEOUT = (isTesting ? 1 : 3) * 1000; // msec
export const SEND_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const TERM_TIMEOUT = (isTesting ? 1 : 10) * 1000; // msec
export const SCAN_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
export const CONNECTION_TIMEOUT = (isTesting ? 1 : 60) * 1000; // msec
