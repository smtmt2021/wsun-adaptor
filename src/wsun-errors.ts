/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export class WsunError extends Error {
  constructor(message: string) {
    super('wsun: ' + message);
    this.name = new.target.name;
  }
}

export class WsunUdpTransmissionFailed extends WsunError {
  constructor() {
    super("wsun adaptor replied 'udp transmission failed'");
    this.name = new.target.name;
  }
}

export class WsunSendingInProgress extends WsunError {
  constructor() {
    super('wsun is sending in progress');
    this.name = new.target.name;
  }
}

export class WsunCommandError extends WsunError {
  private static reason2message(reason: number): string {
    switch (reason) {
      case 4:
        return "wsun adaptor replied 'unsupported command'";
      case 5:
        return "wsun adaptor replied 'incorrect command'";
      case 6:
        return "wsun adaptor replied 'incorrect range'";
      case 9:
        return "wsun adaptor replied 'uart error'";
      case 10:
        return "wsun adaptor replied 'failed to get result'";
      default:
        return "wsun adaptor replied 'unknown error'";
    }
  }

  constructor(public reason: number) {
    super(WsunCommandError.reason2message(reason));
    this.name = new.target.name;
  }
}

export class WsunTimeoutError extends WsunError {
  constructor(message: string) {
    super(message);
    this.name = new.target.name;
  }
}
