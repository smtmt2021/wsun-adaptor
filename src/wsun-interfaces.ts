/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { RemoteInfo } from 'dgram';

export type WsunEventUdp = 'rxudp';

export interface WsunCache {
  channel: number;
  panId: number;
  addr: string;
}

export interface WsunPanDesc extends WsunCache {
  page?: number;
  pairId?: string;
  LQI?: number;
}

export interface WsunId {
  id: string;
  password: string;
}

export interface WsunUDP {
  sender: string;
  dest: string;
  rport: number;
  lport: number;
  senderlla: string;
  secured: boolean;
  datalen: number;
  data: Buffer;
}

export interface WsunSocket {
  bind(port: number): Promise<void>;
  close(): Promise<void>;
  send(msg: Buffer | string, port: number, address: string): Promise<void>;
  on(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  once(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeListener(event: 'message', listener: (msg: Buffer, rinfo: RemoteInfo) => void): this;
  removeAllListeners(event: 'message'): this;
}

export interface WsunConfig {
  device: string; // device name of serial port which is connected with wsun adaptor
}

export interface WsunAdaptorInternal {
  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Delete a used socket.
   *  This function is called automatically whenever a socket is closed.
   *  @param  socket  to be delete.
   */
  deleteSocket(socket: WsunSocket): void;

  /**
   *  Send a datagram message to the specified destination.
   *  This function is called via socket, do not call directly.
   *  @param  msg  to be sent.
   *  @param  port  is destination port.
   *  @param  address is destination address.
   */
  sendDgram(msg: Buffer, port: number, address: string): Promise<void>;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'rxudp'
   */
  on(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;

  /**
   * removes the specified listener from the event
   * @param en
   * @param listener
   */
  removeListener(en: WsunEventUdp, listener: (udp: WsunUDP) => void): void;
}
