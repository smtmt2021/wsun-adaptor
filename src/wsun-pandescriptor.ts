/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunPanDesc } from './';

export class PanDescriptor {
  private descriptor = {} as WsunPanDesc;

  reset(): void {
    this.descriptor = {} as WsunPanDesc;
  }

  get isValid(): boolean {
    return !(this.descriptor.channel == null
      || this.descriptor.page == null
      || this.descriptor.panId == null
      || this.descriptor.addr == null
      || this.descriptor.LQI == null
      || this.descriptor.pairId == null
    );
  }

  set channel(arg: string) {
    const channel = parseInt(arg, 16);
    if (!isNaN(channel)) this.descriptor.channel = channel;
  }

  set page(arg: string) {
    const page = parseInt(arg, 16);
    if (!isNaN(page)) this.descriptor.page = page;
  }

  set panId(arg: string) {
    const panId = parseInt(arg, 16);
    if (!isNaN(panId)) this.descriptor.panId = panId;
  }

  set addr(arg: string) {
    this.descriptor.addr = arg;
  }

  set LQI(arg: string) {
    const LQI = parseInt(arg, 16);
    if (!isNaN(LQI)) this.descriptor.LQI = LQI;
  }

  set pairId(arg: string) {
    this.descriptor.pairId = arg;
  }

  getDescriptor(): WsunPanDesc {
    const descriptor = Object.assign({}, this.descriptor);
    this.reset();
    return descriptor;
  }

  toString(): string {
    return JSON.stringify(this.descriptor);
  }
}
