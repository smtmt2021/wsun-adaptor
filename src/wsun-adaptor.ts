/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunId, WsunPanDesc, WsunSocket } from './';
import { WsunCache } from './wsun-interfaces';

export type WsunEventSession =
  | 'session-established'
  | 'session-ending'
  | 'session-ended'
  | 'session-error'
  | 'session-timeout'
  | 'session-108-start'
  | 'session-108-end';

export interface WsunAdaptor {
  /**
   *  Start activities with a wsun adaptor
   *  and establish a connection with PAA coordinator.
   *  @param  wsunId  ID & password
   *  @param  cache   cache info
   *  @returns  a descriptor of the connected PAA coordinator,
   *            undefined if no PAA coordinator was found.
   */
  open(wsunId: WsunId, cache?: WsunCache): Promise<WsunPanDesc | undefined>;

  /**
   *  Terminate a connection with a PAA coordinator
   */
  close(): Promise<void>;

  /**
   *  Check if connection with wsun adaptor.
   *  @return true if the connection is online.
   */
  isOnline(): boolean;

  /**
   *  Create a socket to communicate remote node.
   *  @return a socket.
   */
  createSocket(): WsunSocket;

  /**
   * Add a listener for the event
   * @param en
   * @param listener
   * @event 'session-established' a PANA session is established
   * @event 'session-ending'      receive a PANA session end request
   * @event 'session-ended'       a PANA session is ended
   * @event 'session-error'       a PANA session is ended due to error
   * @event 'session-timeout'     a PANA session is ended due to timeout
   * @event 'session-108-start'   Restriction of ARIB 108 is started
   * @event 'session-108-end'     Restriction of ARIB 108 is ended
   */
  on(en: WsunEventSession, listener: () => void): void;
}
