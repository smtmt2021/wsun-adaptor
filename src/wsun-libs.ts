/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunCache } from './wsun-interfaces';

export const isParamOfPanDesc = /^ {2}[A-Za-z0-9 ]+:[A-Fa-f0-9]+$/;
export const isNumber = /^\d+$/;
export const isMac = /^[0-9A-Fa-f]{16}$/;
export const isIPv6 = /^[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+:[0-9A-F]+$/;

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function msec2duration(msec: number): number {
  msec = Math.max(msec, 20);
  const duration = Math.floor(Math.log2(msec / 10) - 1);
  return Math.min(duration, 14);
}

export function duration2msec(duration: number): number {
  return Math.pow(2, duration + 1) * 10;
}

export function isValidCache(cache: WsunCache): boolean {
  if (cache.channel < 33 || cache.channel > 60) {
    return false;
  }
  if (cache.panId < 0 || cache.panId > 0xffff) {
    return false;
  }
  if (!isMac.test(cache.addr)) {
    return false;
  }
  return true;
}
