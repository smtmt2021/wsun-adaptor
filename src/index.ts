/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { WsunConfig } from './wsun-interfaces';
import { WsunAdaptor, WsunEventSession } from './wsun-adaptor';
import { WsunBp35a1 } from './wsun-bp35a1';
import DEBUG from 'debug';
const debug = DEBUG('wsun/adaptor');

export { MAX_DGRAM_SIZE } from './wsun-device-constants';
export { WsunAdaptor, WsunEventSession };
export { WsunId, WsunPanDesc, WsunSocket, WsunConfig, WsunCache } from './wsun-interfaces';
export {
  WsunError,
  WsunSendingInProgress,
  WsunUdpTransmissionFailed,
  WsunCommandError,
  WsunTimeoutError
} from './wsun-errors';

export type NameOfClass = 'BP35A1' | 'MOCK';
export interface MockConstructor {
  new (config: WsunConfig): WsunAdaptor;
}

let instance: WsunAdaptor;

/**
 *  Class factory to get a wsun adaptor
 *  @param  className is model name of adaptor you want
 *  @param  config
 *  @return an instance of adaptor which is specified by the model name
 */
export function WsunGetAdaptor(
  className: NameOfClass,
  config?: WsunConfig
): WsunAdaptor | undefined {
  debug('WsunGetAdaptor:', className, JSON.stringify(config));
  if (!instance && config) {
    switch (className) {
      case 'BP35A1':
        instance = new WsunBp35a1(config);
        break;
    }
  }
  return instance;
}
