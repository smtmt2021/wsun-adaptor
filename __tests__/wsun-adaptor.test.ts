/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Bp35a1Robot as Robot, MockBinding, SerialPort } from '@smtmt2021/wsun-adaptor-mock';
import { WsunGetAdaptor, WsunAdaptor } from '../src';
import * as WDC from '../src/wsun-device-constants';
import { WsunUDP, WsunAdaptorInternal } from '../src/wsun-interfaces';

const FAKE_PORT = '/dev/ROBOT';
const FAKE_ID = '0123456789ABCDEF0123456789ABCDEF';
const FAKE_PASSWORD = 'ABDEFGHIJKLM';
const REMOTE_DEVICE_MAC = '123ABC456DEF7890';
const REMOTE_DEVICE2_MAC = '123ABC456DEF7892';
const REMOTE_DEVICE_IPV6 = 'FE80:0000:0000:0000:103A:BC45:6DEF:7890';
const REMOTE_DEVICE2_IPV6 = 'FE80:0000:0000:0000:103A:BC45:6DEF:7892';
const MY_DEVICE_IPV6 = '2001:0DB8:0000:0000:0000:0000:0000:0001';

SerialPort.Binding = Robot;
MockBinding.createPort(FAKE_PORT, { echo: false, record: true });

type WsunAdaptorTest = WsunAdaptor & WsunAdaptorInternal;
let wsAdaptor: WsunAdaptorTest;

function sleep(msec: number): Promise<void> {
  return new Promise<void>(resolve => setTimeout(resolve, msec));
}

describe('WSunAdaptor', () => {
  beforeAll(async () => {
    console.error = function () {
      // disable annoying error message
    };
    jest.setTimeout(10 * 1000);
    wsAdaptor = WsunGetAdaptor('BP35A1', {
      device: FAKE_PORT
    }) as WsunAdaptorTest;
    Robot.instance.panDescriptor.addr = REMOTE_DEVICE_MAC;
    Robot.instance.ipv6 = REMOTE_DEVICE_IPV6;
    await sleep(1);
  });

  beforeEach(() => {
    Robot.instance.udps = [];
    Robot.instance.noCommandReply = false;
    Robot.instance.noScanReply = false;
    Robot.instance.noScanDevice = false;
    Robot.instance.noJoinReply = false;
    Robot.instance.noTermReply = false;
    Robot.instance.noLookupReply = false;
    Robot.instance.noOkParam = false;
    Robot.instance.failJoin = false;
    Robot.instance.termReplyEvent28 = false;
    Robot.instance.termFailEr10 = false;
    Robot.instance.udpRetryCount = 0;
    Robot.instance.lookupTable = {};
    Robot.instance.lookupTable[REMOTE_DEVICE_MAC] = REMOTE_DEVICE_IPV6;
  });

  test('open failed due to invalid id', () => {
    const wsid = {
      id: FAKE_ID + 'ABC',
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open failed due to invalid password', () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD + '123456789012345678901234567890123456789'
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open failed due to invalid cache', async () => {
    await expect(
      wsAdaptor.open(
        {
          id: FAKE_ID,
          password: FAKE_PASSWORD
        },
        {
          channel: 0, // <<<
          panId: 1,
          addr: '0011223344556677'
        }
      )
    ).rejects.toMatchSnapshot();

    await expect(
      wsAdaptor.open(
        {
          id: FAKE_ID,
          password: FAKE_PASSWORD
        },
        {
          channel: 33,
          panId: 0x10000, // <<<
          addr: '0011223344556677'
        }
      )
    ).rejects.toMatchSnapshot();

    await expect(
      wsAdaptor.open(
        {
          id: FAKE_ID,
          password: FAKE_PASSWORD
        },
        {
          channel: 33,
          panId: 1,
          addr: '00112233445566GG' // <<<<
        }
      )
    ).rejects.toMatchSnapshot();
  });

  test('open throw SEND TIMEOUT', () => {
    Robot.instance.noCommandReply = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open failed due to no scan complete event', () => {
    Robot.instance.noScanReply = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open failed due to no device while scanning', () => {
    Robot.instance.noScanDevice = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).resolves.toBeUndefined();
  });

  test('open failed due to no join device', () => {
    Robot.instance.noJoinReply = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open failed due to event24', () => {
    Robot.instance.failJoin = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).resolves.toBeUndefined();
  });

  test('open failed due to lookup', () => {
    Robot.instance.noLookupReply = true;
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('open successfully', () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    const sessionEstablished = jest.fn();
    wsAdaptor.on('session-established', sessionEstablished);

    return wsAdaptor.open(wsid).then(pandesc => {
      expect(wsAdaptor.isOnline()).toBe(true);
      expect(Robot.instance.panDescriptor).toEqual(pandesc);
      expect(Robot.instance.panDescriptor.channel).toBe(Robot.instance.reg.S02);
      expect(Robot.instance.panDescriptor.panId).toBe(Robot.instance.reg.S03);
      expect(Robot.instance.rbid).toBe(wsid.id);
      expect(Robot.instance.password).toBe(wsid.password);
      expect(Robot.instance.reg.SFE).toBe(0);
      expect(Robot.instance.opt).toBe(1);
      expect(sessionEstablished).toHaveBeenCalled();
      // wsAdaptor.removeAllListeners('session-established');
    });
  });

  test('re-open throw INVALD STATE', () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return expect(wsAdaptor.open(wsid)).rejects.toMatchSnapshot();
  });

  test('send data', async () => {
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    await expect(wsAdaptor.sendDgram(data, port, dest)).resolves.not.toThrow();
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(data);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('send data to mac', async () => {
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_MAC;
    await expect(wsAdaptor.sendDgram(data, port, dest)).resolves.not.toThrow();
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(data);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('send data to invalid port', async () => {
    const data = Buffer.from('01234567890123456789');
    const port = 0x10000;
    const dest = REMOTE_DEVICE_IPV6;
    await expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
  });

  test('send large data', () => {
    const data = Buffer.alloc(WDC.MAX_DGRAM_SIZE + 1);
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    return expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
  });

  test('send data to invalid address', () => {
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = '0123456789ABCDEG';
    return expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
  });

  test('send data (retry: 1)', async () => {
    Robot.instance.udpRetryCount = 1;
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    await wsAdaptor.sendDgram(data, port, dest);
    expect(Robot.instance.udpRetryCount).toBe(0);
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(data);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('send data (retry: 2)', async () => {
    Robot.instance.udpRetryCount = 2;
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    await wsAdaptor.sendDgram(data, port, dest);
    expect(Robot.instance.udpRetryCount).toBe(0);
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(data);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('send data (retry: 3)', async () => {
    Robot.instance.udpRetryCount = 3;
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    await wsAdaptor.sendDgram(data, port, dest);
    expect(Robot.instance.udpRetryCount).toBe(0);
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(data);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('send data (retry: 4)', async () => {
    Robot.instance.udpRetryCount = 4;
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    await expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
    expect(Robot.instance.udpRetryCount).toBe(0);
  });

  test('send data (no handle to send)', async () => {
    const data = Buffer.from('01234567890123456789');
    const port = 1000;
    const dest = REMOTE_DEVICE_IPV6;
    await expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
  });

  test('send data simultaneously to make an exception', async () => {
    const buf1 = Buffer.from('this is a test#1');
    const buf2 = Buffer.from('this is a test#2');
    const port = 3610;
    const dest = REMOTE_DEVICE_IPV6;
    const promise = wsAdaptor.sendDgram(buf1, port, dest);
    await expect(wsAdaptor.sendDgram(buf2, port, dest)).rejects.toMatchSnapshot();
    await expect(promise).resolves.not.toThrow();
    expect(Robot.instance.udps.length).toBe(1);
    expect(Robot.instance.udps[0].data).toEqual(buf1);
    expect(Robot.instance.udps[0].ipaddr).toBe(REMOTE_DEVICE_IPV6);
    expect(Robot.instance.udps[0].port).toBe(port);
  });

  test('receive data', done => {
    const dest = MY_DEVICE_IPV6;
    const rport = 3610;
    const lport = 1000;
    const data = Buffer.from('this is a test data');
    wsAdaptor.on('rxudp', (udp: WsunUDP) => {
      expect(udp.data).toEqual(data);
      expect(udp.datalen).toBe(data.length);
      expect(udp.sender).toBe(Robot.instance.ipv6);
      expect(udp.senderlla).toBe(REMOTE_DEVICE_MAC);
      expect(udp.dest).toBe(dest);
      expect(udp.lport).toBe(lport);
      expect(udp.rport).toBe(rport);
      expect(udp.secured).toBeTruthy();
      done();
    });
    Robot.instance.sendBack(dest, rport, lport, data);
  });

  test('fire session events', async () => {
    const sessionTimeout = jest.fn();
    wsAdaptor.on('session-timeout', sessionTimeout);
    Robot.instance.issueEvent(0x29);
    await sleep(100);
    expect(sessionTimeout).toHaveBeenCalled();

    const session108start = jest.fn();
    wsAdaptor.on('session-108-start', session108start);
    Robot.instance.issueEvent(0x32);
    await sleep(100);
    expect(session108start).toHaveBeenCalled();

    const session108end = jest.fn();
    wsAdaptor.on('session-108-end', session108end);
    Robot.instance.issueEvent(0x33);
    await sleep(100);
    expect(session108end).toHaveBeenCalled();
  });

  test('close WSun adaptor', async () => {
    const sessionEnded = jest.fn();
    wsAdaptor.on('session-ended', sessionEnded);

    await expect(wsAdaptor.close()).resolves.not.toThrow();
    expect(sessionEnded).toHaveBeenCalled();
    // wsAdaptor.removeAllListeners('session-ended');
  });

  test('close WSun adaptor, again', () => {
    return expect(wsAdaptor.close()).resolves.not.toThrow();
  });

  test('send data when not online', () => {
    const data = Buffer.from('01234567890123456789');
    const port = 3610;
    const dest = '0123456789ABCDEF';
    return expect(wsAdaptor.sendDgram(data, port, dest)).rejects.toMatchSnapshot();
  });

  test('open with cached descriptor successfully, and close successfully', async () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    const cache = {
      channel: 59,
      panId: 4068,
      addr: REMOTE_DEVICE2_MAC
    };
    const pandesc = await wsAdaptor.open(wsid, cache);
    expect(pandesc != null).toBeTruthy();
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(pandesc!.channel).toEqual(cache.channel);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(pandesc!.panId).toEqual(cache.panId);
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    expect(pandesc!.addr).toEqual(cache.addr);
    expect(wsAdaptor.isOnline()).toBe(true);
    expect(cache.channel).toBe(Robot.instance.reg.S02);
    expect(cache.panId).toBe(Robot.instance.reg.S03);
    expect(Robot.instance.rbid).toBe(wsid.id);
    expect(Robot.instance.password).toBe(wsid.password);
    expect(Robot.instance.reg.SFE).toBe(0);
    expect(Robot.instance.opt).toBe(1);
    expect(Robot.instance.lookupTable[REMOTE_DEVICE2_MAC]).toBe(REMOTE_DEVICE2_IPV6);

    await expect(wsAdaptor.close()).resolves.not.toThrow();
  });

  test('open successfully, and close successfully by event28', async () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    const pandesc = await wsAdaptor.open(wsid);
    expect(pandesc != null).toBeTruthy();
    expect(wsAdaptor.isOnline()).toBe(true);
    expect(Robot.instance.panDescriptor).toEqual(pandesc);
    expect(Robot.instance.panDescriptor.channel).toBe(Robot.instance.reg.S02);
    expect(Robot.instance.panDescriptor.panId).toBe(Robot.instance.reg.S03);
    expect(Robot.instance.rbid).toBe(wsid.id);
    expect(Robot.instance.password).toBe(wsid.password);
    expect(Robot.instance.reg.SFE).toBe(0);
    expect(Robot.instance.opt).toBe(1);

    Robot.instance.termReplyEvent28 = true;
    await expect(wsAdaptor.close()).resolves.not.toThrow();
  });

  test('open successfully, get event26, and close successfully', () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    return wsAdaptor
      .open(wsid)
      .then(pandesc => {
        expect(pandesc != null).toBeTruthy();
        expect(wsAdaptor.isOnline()).toBe(true);
        expect(Robot.instance.panDescriptor).toEqual(pandesc);
        expect(Robot.instance.panDescriptor.channel).toBe(Robot.instance.reg.S02);
        expect(Robot.instance.panDescriptor.panId).toBe(Robot.instance.reg.S03);
        expect(Robot.instance.rbid).toBe(wsid.id);
        expect(Robot.instance.password).toBe(wsid.password);
        expect(Robot.instance.reg.SFE).toBe(0);
        expect(Robot.instance.opt).toBe(1);
      })
      .then(() => {
        return new Promise<void>(resolve => {
          wsAdaptor.on('session-ending', resolve);
          Robot.instance.issueEvent(0x26);
        });
      })
      .then(() => expect(wsAdaptor.close()).resolves.not.toThrow())
      .then(() => expect(wsAdaptor.isOnline()).toBeFalsy());
  });

  test('open successfully, and close erroneously due to term-timeout', async () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    const pandesc = await wsAdaptor.open(wsid);
    expect(pandesc != null).toBeTruthy();
    expect(Robot.instance.panDescriptor).toEqual(pandesc);
    expect(Robot.instance.panDescriptor.channel).toBe(Robot.instance.reg.S02);
    expect(Robot.instance.panDescriptor.panId).toBe(Robot.instance.reg.S03);
    expect(Robot.instance.rbid).toBe(wsid.id);
    expect(Robot.instance.password).toBe(wsid.password);
    expect(Robot.instance.reg.SFE).toBe(0);
    expect(Robot.instance.opt).toBe(1);

    Robot.instance.noTermReply = true;
    await expect(wsAdaptor.close()).resolves.toBeUndefined();
  });

  test('open successfully, and close erroneously due to fail er10', async () => {
    const wsid = {
      id: FAKE_ID,
      password: FAKE_PASSWORD
    };
    const pandesc = await wsAdaptor.open(wsid);
    expect(pandesc != null).toBeTruthy();
    expect(Robot.instance.panDescriptor).toEqual(pandesc);
    expect(Robot.instance.panDescriptor.channel).toBe(Robot.instance.reg.S02);
    expect(Robot.instance.panDescriptor.panId).toBe(Robot.instance.reg.S03);
    expect(Robot.instance.rbid).toBe(wsid.id);
    expect(Robot.instance.password).toBe(wsid.password);
    expect(Robot.instance.reg.SFE).toBe(0);
    expect(Robot.instance.opt).toBe(1);

    Robot.instance.termFailEr10 = true;
    await expect(wsAdaptor.close()).resolves.toBeUndefined();
  });
});
