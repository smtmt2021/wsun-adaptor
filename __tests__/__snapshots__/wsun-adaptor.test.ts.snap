// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WSunAdaptor open failed due to invalid cache 1`] = `[TypeError: Spcecified cache is invalid]`;

exports[`WSunAdaptor open failed due to invalid cache 2`] = `[TypeError: Spcecified cache is invalid]`;

exports[`WSunAdaptor open failed due to invalid cache 3`] = `[TypeError: Spcecified cache is invalid]`;

exports[`WSunAdaptor open failed due to invalid id 1`] = `[RangeError: Wsun ID must be 32 charactors]`;

exports[`WSunAdaptor open failed due to invalid password 1`] = `[RangeError: Wsun password must be 1 <= and <= 32]`;

exports[`WSunAdaptor open failed due to lookup 1`] = `[WsunTimeoutError: wsun: command timeout]`;

exports[`WSunAdaptor open failed due to no join device 1`] = `[Error: Timeout while making a PANA connection]`;

exports[`WSunAdaptor open failed due to no scan complete event 1`] = `[Error: Timeout while scanning PAN]`;

exports[`WSunAdaptor open throw SEND TIMEOUT 1`] = `[WsunTimeoutError: wsun: command timeout]`;

exports[`WSunAdaptor re-open throw INVALD STATE 1`] = `[Error: Invalid state]`;

exports[`WSunAdaptor send data (no handle to send) 1`] = `[WsunError: wsun: no handle to send a dgram]`;

exports[`WSunAdaptor send data (retry: 4) 1`] = `[WsunUdpTransmissionFailed: wsun: wsun adaptor replied 'udp transmission failed']`;

exports[`WSunAdaptor send data simultaneously to make an exception 1`] = `[WsunSendingInProgress: wsun: wsun is sending in progress]`;

exports[`WSunAdaptor send data to invalid address 1`] = `[TypeError: '0123456789ABCDEG' must be ipv6 or mac address]`;

exports[`WSunAdaptor send data to invalid port 1`] = `[RangeError: The port muse be 1 <= and <= 0xffff]`;

exports[`WSunAdaptor send data when not online 1`] = `[Error: Invalid state]`;

exports[`WSunAdaptor send large data 1`] = `[Error: Size of dgram is too large, 1025]`;
